tuning/data/data_handlers.py

```diff
@@ -243,20 +249,84 @@ def apply_tokenizer_chat_template(
         element: the HF Dataset element.
         tokenizer: Tokenizer to be used.
         dataset_text_field: formatted_dataset_field.
+        chat_data_key: dataset field where chat template will be applied.
     Returns:
         Formatted HF Dataset element by formatting dataset with tokenizer's chat template
         Saves the result to dataset_text_field argument.
     """
+    processor = kwargs.get("processor", None)
+    if processor is not None:
+        tokenizer = processor
     if tokenizer.chat_template is None:
         raise ValueError(
             "Tokenizer does not contain tokenizer.chat_template\
                           please pass data_args.chat_template"
         )
+    if chat_data_key and chat_data_key in element:
+        element = element[chat_data_key]
+
     return {
         f"{dataset_text_field}": tokenizer.apply_chat_template(element, tokenize=False)
     }
 
 
+def apply_multimodal_data_processor(
+    element: Dict[str, str],
+    processor: Union[AutoProcessor, LlavaProcessor],
+    **kwargs,
+):
+    """Function (data handler) to apply processor to multimodal dataset elements.
+       Expects to be run as a HF Map API function.
+    Args:
+        element: the HF Dataset element.
+        processor: The processor instance of AutoProcessor or LlavaProcessor.
+    Returns:
+        Formatted HF Dataset element by formatting dataset with processor
+    """
+
+    processor_kwargs = kwargs.get("processor_kwargs", {})
+    fields_name = kwargs.get("fields_name", {})
+    try:
+        text_field = fields_name["dataset_text_field"]
+        image_field = fields_name["dataset_image_field"]
+    except KeyError as e:
+        raise ValueError(f"Missing required field in fields_name: {e}") from e
+
+    text = element.get(text_field)
+    image = element.get(image_field)
+
+    if text is None or image is None:
+        raise ValueError("Missing text or image data in element.")
+
+    # Handler is used with batch=True where image is `List[List[PIL.Image], List[PIL.Image]]`
+    # We need to convert it to `List[PIL.Image]` for LlavaProcessor
+    if isinstance(processor, LlavaProcessor):
+        if isinstance(image, list) and image and isinstance(image[0], list):
+            image = [img[0] for img in image]
+        else:
+            raise ValueError(
+                "Expected image to be a non-empty list of lists \
+                with a single image for LlavaProcessor."
+            ) from e
```

Comment:
@anhuong I missed this in my PR :) Thank you.
```suggestion
            )
```